wp/lib/bap_wp/src/z3_utils.ml

```diff
@@ -100,6 +100,113 @@ let get_z3_name (map : Constr.z3_expr Var.Map.t) (name : string) (fmt : Var.t ->
+let make_arg_synonyms
+  ?postfix:(postfix = "")
+  (sub : Sub.t)
+  (init_var_map : Expr.expr Var.Map.t)
+  (var_map : Expr.expr Var.Map.t) : Sexp.t list =
+  let args = Term.enum arg_t sub |> Seq.to_list in
+  List.concat_map args ~f:(fun arg ->
+    let cname = Arg.lhs arg in
+    let syn prefix z3_reg =
+      let sort = Sexp.of_string (Z3.Sort.to_string (Z3.Expr.get_sort z3_reg)) in
+      synonym (prefix ^ Var.name cname ^ postfix) sort
+      (Sexp.Atom  (Expr.to_string z3_reg))
+    in
+    let reg = Arg.rhs arg in
+    match reg with
+    | Var reg -> begin
+         match Arg.intent arg with
+           | Some In   -> [syn "init_" (Var.Map.find_exn init_var_map reg)]
+           | Some Out  -> [syn "" (Var.Map.find_exn var_map reg)]
+           | Some Both -> [syn "init_" (Var.Map.find_exn init_var_map reg);
+                           syn "" (Var.Map.find_exn var_map reg)]
+           | None -> []
+    end
+    | _ -> [] (* We don't make synonym unless c variable is just held in a bap variable *)
+    )
+
+(** [args_prelude_compare] builds smtlib synonyms in a comparison for the arguments of
+    two compared subroutines. *)
+let args_prelude_compare
+  (main_sub1 : Sub.t)
+  (main_sub2 : Sub.t)
+  (env1 : Env.t)
+  (env2 : Env.t) : Sexp.t list =
+  let var_map1 = Env.get_var_map env1 in
+  let var_map2 = Env.get_var_map env2 in
+  let init_var_map1 = Env.get_init_var_map env1 in
+  let init_var_map2 = Env.get_init_var_map env2 in
+  let syn1 = make_arg_synonyms ~postfix:"_orig" main_sub1 init_var_map1 var_map1 in
+  let syn2 = make_arg_synonyms ~postfix:"_mod" main_sub2 init_var_map2 var_map2 in
+  syn1 @ syn2
+
+
+
+(** [def_pointer_sort] constructs an smtlib sort synonym of pointer sort.
+    Typically it will be `(define-sort pointer () BV64)` *)
+let def_pointer_sort env =
+  let mem = Env.get_mem env in
+  let var_map = Env.get_var_map env in
+  let z3_mem = Var.Map.find_exn var_map mem in
+  let pointer_sort = Z3.Z3Array.get_domain (Expr.get_sort z3_mem) in
+  define_sort "pointer" (sexp_of_sort pointer_sort)
+
+(** [def_mem_sort] constructs an smtlib sort synonym of the memory array.
+    Typically it will be `(define-sort memsort () (Array (BV64 BV8))` *)
+let def_mem_sort env =
+    let mem = Env.get_mem env in
+    let var_map = Env.get_var_map env in
+    let z3_mem = Var.Map.find_exn var_map mem in
+    define_sort "memsort" (sexp_of_sort (Z3.Expr.get_sort z3_mem))
+
+let compare_prelude_smtlib2  main_sub1 main_sub2 (env1 : Env.t) (env2 : Env.t) : string =
+ let std_prelude = "
+        (
+        ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+        ; Sort synonyms
+        (define-sort BV8 () (_ BitVec 8))
+        (define-sort BV16 () (_ BitVec 16))
+        (define-sort BV32 () (_ BitVec 32))
+        (define-sort BV64 () (_ BitVec 64))
+        (define-sort char () (_ BitVec 8))
+        (define-sort byte () (_ BitVec 8))
+        (define-sort int16 () (_ BitVec 16))
```

Comment:
What's the difference between a BV16 and an int16 in this case?

Reply:
Nothing. I thought it was reasonable or at least unharmful to have C-esque looking types and something that is just a nice shorthand evocative of (_ BitVec 64). Maybe I should use `int16_t` instead

Reply:
Definitely `int16_t`!